src/include/platform/internal/GenericPlatformManagerImpl_POSIX.cpp

```diff
@@ -257,22 +257,65 @@ CHIP_ERROR GenericPlatformManagerImpl_POSIX<ImplClass>::_StartEventLoopTask()
 }
 
 template <class ImplClass>
-CHIP_ERROR GenericPlatformManagerImpl_POSIX<ImplClass>::_Shutdown()
+CHIP_ERROR GenericPlatformManagerImpl_POSIX<ImplClass>::_StopEventLoopTask()
 {
     int err = 0;
+
+    _LockChipStack();
+
     mShouldRunEventLoop.store(false, std::memory_order_relaxed);
+
     if (mChipTask)
     {
         SystemLayer.WakeSelect();
+
+        _UnlockChipStack();
+
         SuccessOrExit(err = pthread_join(mChipTask, nullptr));
     }
-    // Call up to the base class _Shutdown() to perform the bulk of the shutdown.
-    err = GenericPlatformManagerImpl<ImplClass>::_Shutdown();
+    else
+    {
+        _UnlockChipStack();
+    }
 
 exit:
     return System::MapErrorPOSIX(err);
 }
 
+template <class ImplClass>
+CHIP_ERROR GenericPlatformManagerImpl_POSIX<ImplClass>::_Shutdown()
+{
+    int err = 0;
```

Comment:
Could we remove the goto exit pattern here?

```cpp
int err = _StopEventLoopTask();

if (err != SYSTEM_NO_ERROR) {
    err = GenericPlatformManagerImpl<ImplClass>::_TeardownChipStack();
}

return System::MapErrorPOSIX(err);
```

Maybe if _StopEvent* and _Teardown... return ints,  the _Shutdown should do the same to not even need a MapErrorPOSIX and could use return macros:

```cpp
ReturnErrorOnFailure(_StopEventLoopTask());
return GenericPlatformManagerImpl<ImplClass>::_TeardownChipStack();
```

although I am not sure if the CHIP_ERROR temporary in the macros will work well with the int :/

Reply:
Altering the return values of those functions isn't the scope of this PR (I was trying to keep this very narrow to just fixing the threading issues), so given that, there is a need to do the error conversion.

Also, I think we should have a single return pointer in a given function to permit consolidated 'exit cleanup'?

Reply:
Over time I have found that the 'consolidated exit cleanup' results in code concessions:
  - variables need to be declared at top, hence increasing scope unnecessarely
  - readability-wise a 'return' vs 'ExitNow' is functionally the same with the exit macros incurring extra mental effort (minimal though once you get used to the code)
  - separates code into 'initialization', 'usage' and 'return/cleanup' that for longer functions is harder to read
  
  As a general thing going forward I would want to discourage any usage of 'goto exit' at all. If cleanup is needed, use RAII. The only case I do not have a clean answer for is 'if (err) Log'.
  
  Since this is newly written code, I expect it to not use `goto exit`. It does have somewhat of a cleanup (mapping posix error to chip) but that seems solvable.